Replace SleekXMPP with Slixmpp

Dunedan opened the pull request:
SleekXMPP has been declared end-of-life for years and hasn't seen any
activity since. It also support only Python up to version 3.7. There is
a friendly fork called Slixmpp however, which sees continuous
development and releases and supports newer Python versions. To be able
to support newer Python versions with the lobby bots and to avoid using
abandoned libraries as dependencies, this replaces SleekXMPP with
Slixmpp.

Oceandragon commented on xpartamupp/echelon.py:
```diff
@@ -569,14 +580,16 @@ def _iq_board_list_handler(self, iq):
         command = iq['boardlist']['command']
-        self.leaderboard.get_or_create_player(sleekxmpp.jid.JID(iq['from'], resource="0ad"))
+        jid = JID(iq['from'])
+        jid.resource = "0ad"
```
Would we consider making the form similar to how it's seen previously?

`JID(iq['from']).bare + "/0ad"`

Or maybe higher up we should keep the form where it uses `JID(jid, resource="0ad")`

Dunedan replied:
I believe JIDs should be used in the code as instances of the JID class as much as possible, as that provides more appropriate type information and additional functionality compared to strings. That's why I liked the previous notation (`JID(iq['from'], resource="0ad")`) for generating a JID object out of a JID but at the same time replacing its resource. Unfortunately that's not possible anymore with Slixmpp, as the constructor of the JID class there doesn't have a `resource` parameter anymore. That's why I initialized the JID without specifying anything related the resource and set it afterwards. Your suggestion would result in a string instead of a JID object and adjusting it to return a JID would not only result in the need to initialize two JID objects, but also to still do string concatenation (`JID(JID(iq['from']).bare + "/0ad")`).

Oceandragon reviewed (changes requested):
Reconnecting upon disconnection is not working like this. 
We might like to have a graceful shutdown. Maybe it would be worth catching SIGTERM while we're at it since these are intended to be run as systemd services.

Dunedan commented on xpartamupp/echelon.py:
```diff
@@ -792,18 +807,20 @@ def main():
-    if xmpp.connect((args.xserver, 5222) if args.xserver else None, True, not args.xdisabletls):
-        xmpp.process()
+    if args.xserver:
+        xmpp.connect((args.xserver, 5222))
```
I didn't change the existing logic there and only refactored the code to be more readable. I'd consider a change as you suggest it to be out-of-scope for this PR.

Dunedan commented on xpartamupp/echelon.py:
```diff
@@ -489,25 +497,26 @@ def __init__(self, sjid, password, room, nick, leaderboard):
         self.add_event_handler('muc::%s::got_online' % self.room, self._muc_online)
         self.add_event_handler('muc::%s::got_offline' % self.room, self._muc_offline)
         self.add_event_handler('groupchat_message', self._muc_message)
+        self.add_event_handler('disconnected', lambda _: self.connect())
```
Yes, let's put that into a separate PR though, to keep this one focused on just switching to Slixmpp.

Dunedan commented on xpartamupp/echelon.py:
```diff
@@ -792,18 +807,20 @@ def main():
-        logging.error("Unable to connect")
+        xmpp.connect(None, True, not args.xdisabletls)
+
+    asyncio.get_event_loop().run_forever()
```
Let's handle that in a separate PR to keep this one focused to just switching to Slixmpp.

Dunedan commented on xpartamupp/xpartamupp.py:
```diff
@@ -348,18 +346,20 @@ def main():
-        logging.error("Unable to connect")
+        xmpp.connect(None, True, not args.xdisabletls)
+
+    asyncio.get_event_loop().run_forever()
```
Let's handle that in a separate PR to keep this one focused to just switching to Slixmpp.

Dunedan commented on tests/test_echelon.py:
```diff
@@ -204,26 +205,5 @@ def test_success(self):
             xmpp_mock().connect.assert_called_once_with(None, True, True)
-            xmpp_mock().process.assert_called_once_with()
-
-    def test_failing_connect(self):
```
This test relied on `ClientXMPP().connect()` to return a boolean, which isn't the case anymore with Slixmpp. Therefore this test didn't make any sense anymore.

Dunedan commented on tests/test_xpartamupp.py:
```diff
@@ -162,23 +163,5 @@ def test_success(self):
             xmpp_mock().connect.assert_called_once_with(None, True, True)
-            xmpp_mock().process.assert_called_once_with()
-
-    def test_failing_connect(self):
```
See above.

Dunedan commented on xpartamupp/xpartamupp.py:
```diff
@@ -158,15 +156,16 @@ def __init__(self, sjid, password, room, nick):
         self.add_event_handler('muc::%s::got_online' % self.room, self._muc_online)
         self.add_event_handler('muc::%s::got_offline' % self.room, self._muc_offline)
         self.add_event_handler('groupchat_message', self._muc_message)
+        self.add_event_handler('disconnected', lambda _: self.connect())
```
Good catch. As discussed already I'm not satisfied with the re-connect behavior in general, so I'll have another look at that.

Dunedan replied:
After some debugging I came to the conclusion that calling `self.connect()` after the client got initialized without any arguments is perfectly fine, as Slixmpp is supposed to reuse the parameters provided for the original call of `ClientXMPP.connect()`. However, there is a bug in Slixmpp preventing this to work for some parameters, most notably for the `disable_starttls` parameter you stumbled upon. I already opened a merge request for Slixmpp with a fix (https://lab.louiz.org/poezio/slixmpp/-/merge_requests/226), but this merge request still needs to get accepted into Slixmpp. For the time being I added a workaround so we aren't affected by this bug and also don't need to pass connection parameters explicitly into the XMPP client object.

I also did fix the other points which bothered me regarding the reconnect behavior and it should behave fine now. That means there won't be any re-connection attempts anymore if a connection fails because of invalid credentials and all other cases will use the same back-off logic as the connect functionality of Slixmpp uses internally when certain errors happen during connection attempts.

Oceandragon reviewed (approved):
As these are `breaking changes` until fixes are made to enable the bots to run successfully, it should be merged into its own `feature branch`. Once it's not breaking the bots, it can be merged into master.

Dunedan commented:
If I'm not mistaken the only breaking change is the retry behavior and that's something I'm going to fix before merging this into `master`.

Dunedan commented:
@Oceandragon I addressed the issues with the re-connection behavior now. From my point of view this PR in a state where it can get merged, but I'd appreciate if you'd find the time to have another look at it.
